### PR TITLE
CRI: use more gogoprotobuf plugins

### PR DESCRIPTION
**Pull request by yujuhong**
Generate marshaler/unmarshaler code should help improve the performance.

This addresses #40098

### CODE REVIEW

**Comment by yujuhong** [comment]
/cc @yifan-gu @feiskyer 

This is still WIP. Need to examine all the plugins to see which ones we need.

**Comment by feiskyer** [comment]
A lot of `GetXXX()` methods are removed with those plugins:

```
W0124 18:11:11.969] pkg/kubelet/kuberuntime/helpers.go:143: imageStatus.GetUid undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".Image has no field or method GetUid)
W0124 18:11:11.970] pkg/kubelet/kuberuntime/helpers.go:229: status.GetConditions undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".RuntimeStatus has no field or method GetConditions)
W0124 18:11:11.970] pkg/kubelet/kuberuntime/kuberuntime_container.go:105: containerConfig.GetMetadata undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetMetadata)
W0124 18:11:11.971] pkg/kubelet/kuberuntime/kuberuntime_container.go:106: podSandboxConfig.GetMetadata undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".PodSandboxConfig has no field or method GetMetadata)
W0124 18:11:11.971] pkg/kubelet/kuberuntime/kuberuntime_manager.go:797: podStatus.GetLinux undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".PodSandboxStatus has no field or method GetLinux)
W0124 18:11:12.344] # k8s.io/kubernetes/pkg/kubelet/dockershim
W0124 18:11:12.345] pkg/kubelet/dockershim/docker_container.go:49: filter.GetState undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerFilter has no field or method GetState)
W0124 18:11:12.345] pkg/kubelet/dockershim/docker_container.go:93: config.GetLabels undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetLabels)
W0124 18:11:12.346] pkg/kubelet/dockershim/docker_container.go:93: config.GetAnnotations undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetAnnotations)
W0124 18:11:12.346] pkg/kubelet/dockershim/docker_container.go:102: config.GetImage undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetImage)
W0124 18:11:12.347] pkg/kubelet/dockershim/docker_container.go:111: config.GetEnvs undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetEnvs)
W0124 18:11:12.347] pkg/kubelet/dockershim/docker_container.go:124: config.GetMounts undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetMounts)
W0124 18:11:12.348] pkg/kubelet/dockershim/docker_container.go:128: config.GetLinux undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ContainerConfig has no field or method GetLinux)
W0124 18:11:12.348] pkg/kubelet/dockershim/docker_container.go:150: sandboxConfig.GetLinux undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".PodSandboxConfig has no field or method GetLinux)
W0124 18:11:12.349] pkg/kubelet/dockershim/docker_image.go:31: filter.GetImage undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".ImageFilter has no field or method GetImage)
W0124 18:11:12.349] pkg/kubelet/dockershim/docker_sandbox.go:85: config.GetLinux undefined (type *"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime".PodSandboxConfig has no field or method GetLinux)
```

**Comment by Random-Liu** [comment]
> A lot of GetXXX() methods are removed with those plugins

This should have been fixed in latest commit.

**Comment by feiskyer** [comment]
> This should have been fixed in latest commit.

Sorry, didn't notice this. Only Bazel needs updates.

File: hack/update-generated-runtime-dockerized.sh [hunk 1]
```diff
@@ -46,8 +46,17 @@ function cleanup {
...
+  --proto_path="${KUBE_ROOT}/vendor" \
+  --gogo_out=plugins=grpc:${KUBE_REMOTE_RUNTIME_ROOT} ${KUBE_REMOTE_RUNTIME_ROOT}/api.proto
+
+# Update boilerplate for the generated file.
 echo "$(cat hack/boilerplate/boilerplate.go.txt ${KUBE_REMOTE_RUNTIME_ROOT}/api.pb.go)" > ${KUBE_REMOTE_RUNTIME_ROOT}/api.pb.go
 sed -i".bak" "s/Copyright YEAR/Copyright $(date '+%Y')/g" ${KUBE_REMOTE_RUNTIME_ROOT}/api.pb.go
 
+# Run gofmt to clean up the generated code.
+kube::golang::verify_go_version
+gofmt -l -s -w $KUBE_REMOTE_RUNTIME_ROOT/api.pb.go
```
Comment (Random-Liu) [inline hack/update-generated-runtime-dockerized.sh]: `${KUBE_REMOTE_RUNTIME_ROOT}`

**Review by Random-Liu — COMMENTED** [review]
LGTM with one nit

File: hack/update-generated-runtime-dockerized.sh (hunk 1 above)
Comment (yujuhong) [inline hack/update-generated-runtime-dockerized.sh]: Fixed

**Comment by Random-Liu** [comment]
```
W0125 14:13:53.803] E0125 14:13:53.802889   15684 run_remote.go:478] Failed to collect serial output from node "tmp-node-e2e-c5afc288-e2e-node-ubuntu-trusty-docker12-v1-image": googleapi: Error 503: TIMEOUT, backendError
I0125 14:13:54.039] 
I0125 14:13:54.039] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
I0125 14:13:54.039] >                              START TEST                                >
I0125 14:13:54.039] >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
I0125 14:13:54.039] Start Test Suite on Host tmp-node-e2e-c5afc288-e2e-node-ubuntu-trusty-docker12-v1-image
I0125 14:13:54.039] 
I0125 14:13:54.039] Failure Finished Test Suite on Host tmp-node-e2e-c5afc288-e2e-node-ubuntu-trusty-docker12-v1-image
I0125 14:13:54.040] failed to create workspace directory "/tmp/node-e2e-20170125T141136" on host "tmp-node-e2e-c5afc288-e2e-node-ubuntu-trusty-docker12-v1-image": command [ssh -i /root/.ssh/google_compute_engine -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o CheckHostIP=no -o StrictHostKeyChecking=no -o ServerAliveInterval=30 -o LogLevel=ERROR jenkins@104.197.173.104 -- mkdir /tmp/node-e2e-20170125T141136] failed with error: exit status 255 output: "ssh: connect to host 104.197.173.104 port 22: Connection timed out\r\n"
```

Seems like a GCE flake.
@k8s-bot node e2e test his.

**Comment by yujuhong** [comment]
@k8s-bot node e2e test this

**Comment by yujuhong** [comment]
@k8s-bot gci gke e2e test this

**Comment by yujuhong** [comment]
Tests passed. PTAL.

**Comment by feiskyer** [comment]
LGTM. @Random-Liu  Did you benchmark the performance based on this PR?

**Comment by Random-Liu** [comment]
@feiskyer Not yet. Will do that later. :)

**Comment by yujuhong** [comment]
@spxtr @ixdy, can any one of you review/approve this PR? Thanks!

**Pull request merged**